rest/src/main/scala/org/apache/gearpump/services/AppMasterService.scala

```diff
@@ -18,54 +18,52 @@
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success, Try}
 
-@Api(value = "/appmaster", description = "AppMaster Info.")
-class AppMasterServiceActor(val master:ActorRef) extends Actor with AppMasterService   {
-
-
-  def actorRefFactory = context
-  def receive = runRoute(readRoute)
-}
-
-@Api(value = "/appmaster", description = "AppMaster Info.")
-trait AppMasterService extends HttpService {
+trait AppMasterService extends HttpService  {
   import org.apache.gearpump.services.AppMasterProtocol._
   import spray.httpx.SprayJsonSupport._
-
-  implicit val ec: ExecutionContext = actorRefFactory.dispatcher
-  implicit val timeout = Constants.FUTURE_TIMEOUT
   val master:ActorRef
 
-  val routes = readRoute 
+  def dag(config: Option[UserConfig]): DAG
 
-  @ApiOperation(value = "Get AppMaster Info", notes = "Returns AppMaster Info ", httpMethod = "GET", response = classOf[AppMasterData])
-  @ApiImplicitParams(Array(
-      new ApiImplicitParam(name = "appId", required = true, dataType = "integer", paramType = "path", value = "ID of AppMaster"),
-      new ApiImplicitParam(name = "detail", required = false, dataType = "boolean", paramType = "query", value = "true/false")
-  ))
-  @ApiResponses(Array(
-    new ApiResponse(code = 404, message = "AppMaster not found"),
-    new ApiResponse(code = 400, message = "Invalid ID supplied")
-  ))
-  def readRoute: routing.Route = get {
-     path("appmaster"/IntNumber) { appId => {
+  def appMasterRoute = get {
+    implicit val ec: ExecutionContext = actorRefFactory.dispatcher
+    implicit val timeout = Constants.FUTURE_TIMEOUT
+    path("appmaster"/IntNumber) { appId => {
        parameter("detail" ? "false") { detail =>
          val detailValue = Try(detail.toBoolean).getOrElse(false)
-         onComplete((master ? AppMasterDataRequest(appId, detailValue))
-           .asInstanceOf[Future[AppMasterData]]) {
-           case Success(value: AppMasterData) => complete(value)
-           case Failure(ex) => complete(StatusCodes.InternalServerError, s"An error occurred: ${ex.getMessage}")
+         detailValue match {
+           case true =>
+             onComplete((master ? AppMasterDataDetailRequest(appId)).asInstanceOf[Future[AppMasterDataDetail]]) {
+               case Success(value: AppMasterDataDetail) =>
+                 import upickle._
+                 val application = value.application
+                 Option(application) match {
+                   case Some(app) =>
+                     complete(write(dag(Option(application.userConfig))))
+                   case None =>
+                     complete(StatusCodes.InternalServerError, "UserConfig is null")
+                 }
+               case Failure(ex) => complete(StatusCodes.InternalServerError, s"An error occurred: ${ex.getMessage}")
+             }
+           case false =>
+             onComplete((master ? AppMasterDataRequest(appId)).asInstanceOf[Future[AppMasterData]]) {
```

Comment:
Since you are using uPickle for AppMasterDataDetail, maybe we should use uPickle consistently for all response of REST, like AppMasterData.